File: lib/src/main/java/com/asyncanimator/launcher/async/AsyncValueAnimator.java
package com.asyncanimator.launcher.async;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.ValueAnimator;

import java.util.concurrent.atomic.AtomicBoolean;

/**
 * AsyncValueAnimator — 跨 Looper 安全的 ValueAnimator。
 *
 * <p>对应分析文档 §6.3。start/cancel/end 先判断"当前线程 vs 目标 Looper"，
 * 不一致时通过 LooperExecutor marshal 过去。
 *
 * <p>listener 跨线程派发（mAsyncAnimCallbacks）：listener fire 时再 marshal 回主线程。
 * 平台 {@code AnimatorListener} 的参数是 {@code @NonNull}（Kotlin 侧 override 必须声明非空），
 * 因此派发时传递真实 animator，不再传 null。
 */
public class AsyncValueAnimator extends ValueAnimator {

    private LooperExecutor mAnimLooperExecutor;
    private AsyncAnimCallbacks mAsyncAnimCallbacks;
    private final AtomicBoolean mIsEnd = new AtomicBoolean(false);

    public AsyncValueAnimator() {
        mAnimLooperExecutor = Executors.MAIN_EXECUTOR;
        mAsyncAnimCallbacks = new AsyncAnimCallbacks();
        addListener(new AnimatorListenerAdapter() {
            @Override public void onAnimationCancel(Animator a) {
                if (mIsEnd.get()) return;
                mAsyncAnimCallbacks.onAnimationCancel(a);
            }
            @Override public void onAnimationEnd(Animator a) {
                if (mIsEnd.compareAndSet(false, true)) {
                    mAsyncAnimCallbacks.onAnimationEnd(a);
                }
            }
            @Override public void onAnimationStart(Animator a) {
                if (mIsEnd.get()) return;
                mAsyncAnimCallbacks.onAnimationStart(a);
            }
        });
    }

    public void setExecutor(LooperExecutor exec) {
        mAnimLooperExecutor = exec;
    }

    public AsyncAnimCallbacks getAsyncAnimCallbacks() {
        return mAsyncAnimCallbacks;
    }

    @Override
    public void start() {
        if (isCurrentExecutor()) super.start();
        else mAnimLooperExecutor.execute(super::start);
    }

    @Override
    public void cancel() {
        if (isCurrentExecutor()) super.cancel();
        else mAnimLooperExecutor.execute(super::cancel);
    }

    @Override
    public void end() {
        if (isCurrentExecutor()) super.end();
        else mAnimLooperExecutor.execute(super::end);
    }

    private boolean isCurrentExecutor() {
        return mAnimLooperExecutor.isCurrentThread();
    }
}

File: lib/src/main/java/com/asyncanimator/launcher/async/AsyncValueAnimator.kt
package com.asyncanimator.launcher.async

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.ValueAnimator
import java.util.concurrent.atomic.AtomicBoolean

/**
 * AsyncValueAnimator — 跨 Looper 安全的 ValueAnimator。
 *
 * 对应分析文档 §6.3。start/cancel/end 先判断"当前线程 vs 目标 Looper"，
 * 不一致时通过 LooperExecutor marshal 过去。
 *
 * listener 跨线程派发（[asyncAnimCallbacks]）：listener fire 时再 marshal 回主线程。
 */
class AsyncValueAnimator : ValueAnimator() {

    /** 动画执行器：start/cancel/end 与帧推进所在的 Looper。 */
    var executor: LooperExecutor = Executors.MAIN_EXECUTOR

    /** listener 容器 + 跨线程派发器（listener 始终回主线程 fire）。 */
    val asyncAnimCallbacks = AsyncAnimCallbacks()

    private val isEnd = AtomicBoolean(false)

    init {
        addListener(object : AnimatorListenerAdapter() {
            override fun onAnimationCancel(a: Animator) {
                if (isEnd.get()) return
                asyncAnimCallbacks.onAnimationCancel(a)
            }

            override fun onAnimationEnd(a: Animator) {
                if (isEnd.compareAndSet(false, true)) {
                    asyncAnimCallbacks.onAnimationEnd(a)
                }
            }

            override fun onAnimationStart(a: Animator) {
                if (isEnd.get()) return
                asyncAnimCallbacks.onAnimationStart(a)
            }
        })
    }

    private val isCurrentExecutor: Boolean get() = executor.isCurrentThread

    override fun start() {
        if (isCurrentExecutor) super.start()
        else executor.execute { super@AsyncValueAnimator.start() }
    }

    override fun cancel() {
        if (isCurrentExecutor) super.cancel()
        else executor.execute { super@AsyncValueAnimator.cancel() }
    }

    override fun end() {
        if (isCurrentExecutor) super.end()
        else executor.execute { super@AsyncValueAnimator.end() }
    }
}
